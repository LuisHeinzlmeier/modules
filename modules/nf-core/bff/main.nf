process BFF {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/bioconductor-cellhashr_r-seurat:1c94360d8ed188c4':
        'community.wave.seqera.io/library/bioconductor-cellhashr_r-seurat:25c4bc76749af5ac' }"

    input:
    tuple val(meta), path(hto_matrix), val(methods), val(preprocessing)

    output:
    tuple val(meta), path("*_assignment_bff.csv")       , emit: assignment
    tuple val(meta), path("*_metrics_bff.csv")          , emit: metrics
    tuple val(meta), path("*_params_bff.csv")           , emit: params
    path "versions.yml"                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    methodsForConsensus = task.ext.methodsForConsensus ?: "NULL" // NULL, RAW OR CLUSTER
    cellbarcodeWhitelist = task.ext.cellbarcodeWhitelist ?: "NULL"
    prefix = task.ext.prefix ?: "${meta.id}"

    doTSNE = task.ext.doTSNE ?: "TRUE"
    barcodeWhitelist = task.ext.barcodeWhitelist ?: "NULL"
    doHeatmap = task.ext.doHeatmap ?: "TRUE"
    perCellSaturation = task.ext.perCellSaturation ?: "NULL"
    majorityConsensusThreshold = task.ext.majorityConsensusThreshold ?: "NULL"
    chemistry = task.ext.chemistry ?: "10xV3"
    callerDisagreementThreshold = task.ext.callerDisagreementThreshold ?: "NULL"

    template 'bff.R'

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_assignment_bff.csv
    touch ${prefix}_metrics_bff.csv
    touch ${prefix}_params_bff.csv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(Rscript -e "cat(strsplit(R.version[['version.string']], ' ')[[1]][3])")
        r-seurat: \$(Rscript -e "library(Seurat); cat(as.character(packageVersion('Seurat')))")
        cellhashR: \$(Rscript -e "library(cellhashR); cat(as.character(packageVersion('cellhashR')))")
    END_VERSIONS
    """
}
