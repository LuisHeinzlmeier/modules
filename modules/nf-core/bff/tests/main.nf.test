nextflow_process {

    name "Test Process BFF"
    script "../main.nf"
    process "BFF"

    tag "modules"
    tag "modules_nfcore"
    tag "bff"

    def testDataPath = "/Users/luisheinzlmeier/Desktop/test_data/hto"

    test("hto_matrix - combined - preprocessing") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map
                    "$testDataPath",
                    "COMBINED",
                    "TRUE"

                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("hto_matrix - raw - preprocessing") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map
                    "$testDataPath",
                    "RAW",
                    "TRUE"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("hto_matrix - combined - no preprocessing") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map
                    "$testDataPath",
                    "COMBINED",
                    "FALSE"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("hto_matrix - combined - preprocessing - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map
                    "$testDataPath",
                    "COMBINED",
                    "TRUE"

                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

}
