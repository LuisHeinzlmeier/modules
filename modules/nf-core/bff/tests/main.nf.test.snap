{
    "hto_matrix - combined - preprocessing": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,89808ab6a4f893bf9e59514bd34a3f54"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "3": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,89808ab6a4f893bf9e59514bd34a3f54"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:53:10.676096"
    },
    "hto_matrix - RAW - preprocessing: TRUE": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,7262263d31ceb124df8cbcf3b14f0f07"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,91611097843bd9d7f267927569e78fb0"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,ac4e3948b07098a459574f28b88941b9"
                    ]
                ],
                "3": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,7262263d31ceb124df8cbcf3b14f0f07"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,91611097843bd9d7f267927569e78fb0"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,ac4e3948b07098a459574f28b88941b9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:15:38.72601"
    },
    "hto_matrix - COMBINED - preprocessing: TRUE": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,89808ab6a4f893bf9e59514bd34a3f54"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "3": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,89808ab6a4f893bf9e59514bd34a3f54"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:15:12.425026"
    },
    "hto_matrix - CLUSTER - preprocessing: FALSE": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "assignment": [
                    
                ],
                "metrics": [
                    
                ],
                "params": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:16:09.749385"
    },
    "hto_matrix - COMBINED - preprocessing: FALSE": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,9bdd3fb540ce56b103d424ed833b9239"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "3": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,9bdd3fb540ce56b103d424ed833b9239"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:15:25.298827"
    },
    "hto_matrix - combined - preprocessing - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:53:51.239405"
    },
    "hto_matrix - combined - no preprocessing": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,9bdd3fb540ce56b103d424ed833b9239"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "3": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,9bdd3fb540ce56b103d424ed833b9239"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,761dec35875c035c038fb44a5957eab4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:53:38.441325"
    },
    "sarscov2 - bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "prefix_assignment_bff.csv:md5,89808ab6a4f893bf9e59514bd34a3f54"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "prefix_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "prefix_params_bff.csv:md5,8ea90eadec0e7e3a73800e7f18f0c527"
                    ]
                ],
                "3": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "prefix_assignment_bff.csv:md5,89808ab6a4f893bf9e59514bd34a3f54"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "prefix_metrics_bff.csv:md5,1d7751cf035c9e736e4de5504ff87a6a"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "prefix_params_bff.csv:md5,8ea90eadec0e7e3a73800e7f18f0c527"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T08:42:06.434326"
    },
    "hto_matrix - RAW - preprocessing: FALSE": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,5275e14cb6003ac9048e84df97baaa40"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,91611097843bd9d7f267927569e78fb0"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,ac4e3948b07098a459574f28b88941b9"
                    ]
                ],
                "3": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,5275e14cb6003ac9048e84df97baaa40"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,91611097843bd9d7f267927569e78fb0"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,ac4e3948b07098a459574f28b88941b9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c72f9a10628160cafc48879819ed17ab"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:15:51.245844"
    },
    "hto_matrix - CLUSTER - preprocessing: TRUE": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "assignment": [
                    
                ],
                "metrics": [
                    
                ],
                "params": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:16:00.760488"
    },
    "hto_matrix - raw - preprocessing": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,7262263d31ceb124df8cbcf3b14f0f07"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,91611097843bd9d7f267927569e78fb0"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,ac4e3948b07098a459574f28b88941b9"
                    ]
                ],
                "3": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        "test_assignment_bff.csv:md5,7262263d31ceb124df8cbcf3b14f0f07"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test_metrics_bff.csv:md5,91611097843bd9d7f267927569e78fb0"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_bff.csv:md5,ac4e3948b07098a459574f28b88941b9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,71a50af83aebf56493d64a87f468a6f2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-17T09:53:25.067111"
    }
}